src/container/ParticleContainer.h

```diff
@@ -0,0 +1,161 @@
+#include <string>
+#include <memory>
+#include <spdlog/spdlog.h>
+#include "body/Particle.h"
+#include "body/Cuboid.h"
+#include "force/Force.h"
+
+#pragma once
+
+/**
+ * @brief Define an interface for iteration over particles in a container.
+ */
+class Iterator {
+    /**
+     * The current particle.
+     */
+    std::vector<Particle>::iterator current;
+    /**
+     * The end of the iteration.
+     */
+    std::vector<Particle>::iterator end;
+public:
+    /**
+     * Constructor.
+     * @param begin The first particle.
+     * @param end The end of the iteration.
+     */
+    Iterator(std::vector<Particle>::iterator begin, std::vector<Particle>::iterator end);
+    /**
+     * Destructor.
+     */
+    ~Iterator() = default;
+     /**
+      * Determine the next particle in the container and update the iterator.
+      * @return The next particle in the container
+      */
+    Particle& next();
+     /**
+      * Determine whether there are further particles in the container.
+      * @return True if the end of the container is not yet reached.
+      */
+    bool hasNext();
+};
+
+/**
+ * @brief The abstract class that represents a container for particles. Subclasses implement the concrete way to manage particles.
+ */
+class ParticleContainer {
+
+protected:
+/**
+ * Particles stored in the container
+ */
+    std::unique_ptr<std::vector<Particle>> particles;
```

Comment:
The thing is I don't want to create a empty vector in the main function and pass it to the XML reader, because adding an extra parameter just for passing an empty vector does not seem to be a good idea for me (or is this actually better?). I want to create the vector in the function for reading xml instead. To make sure that the memory for storing particles remains valide during the entire program, there only seem to be two options: either storing the particles directly (not as reference) in the container, which means we have to copy all the particles into it, or storing them on the heap. To do the latter, the only possibilities seem to be creating pointers. Since smart pointers automatically free the memory once they go out of scope, I have to store them as attributes (right?), because if I just store a reference to the vector on the heap, the vector will automatically be destructed.

Reply:
Hi, I saw your comment "unnecessary pointers" in the grading, is it bad to store particles as pointers? Do you have any suggestion on this? Also, you mentioned raw pointers, which I think we didn't use. Could you please point out where it is?

Reply:
Sorry I didn't reply to your comment. I think passing an empty vector is practically free and free like a much nicer way to implement this. I do find use of pointers here a bit unnecessary. About raw pointers, I don't remember if I noticed it somewhere or I made a mistake, in any case, it is nice if you didn't use it :)